pyanaconda/modules/network/nm_client.py

```diff
@@ -51,6 +54,133 @@
+    :rtype: NM.NMClient
+    """
+    if not SystemBus.check_connection():
+        log.debug("get new NM Client failed: SystemBus connection check failed.")
+        return None
+
+    nm_client = NM.Client.new(None)
+
+    if not nm_client.get_nm_running():
+        log.debug("get new NM Client failed: NetworkManager is not running.")
+        return None
+
+    log.debug("get new NM Client succeeded.")
+    return nm_client
+
+
+class NMCallResult():
+    """Result of NM async finish callback."""
+    def __init__(self):
+        self.received_data = None
+        self.error_message = ""
+        self.timeout = False
+
+    @property
+    def succeeded(self):
+        """The async call has succeeded."""
+        return not self.failed
+
+    @property
+    def failed(self):
+        """The async call has failed."""
+        return self.error_message or self.timeout
+
+
+def sync_call_nm_client(context, async_call, async_call_finish, timeout, *args, **kwargs):
+    """Call libnm asynchronous method synchronously with timeout.
+
+    :param context: context for the loop
+    :type context: GMainContext
+    :param async_call: asynchronous NM method to be called
+    :type async_call: libnm method
+    :param async_call_finish: finish method of the asynchronous call
+    :type async_call_finish: libnm method
+    :param timeout: timeout for the loop in seconds (0 == no timeout)
+    :type timeout: int
+
+    *args should hold all positional arguments preceding the cancellable argument
+    """
+
+    info = async_call.get_symbol()
+    result = NMCallResult()
+
+    loop = create_main_loop(context)
+    callbacks = [loop.quit]
+
+    def _stop_loop():
+        log.debug("sync_call_nm_client[%s]: quit", info)
+        while callbacks:
+            callback = callbacks.pop()
+            callback()
+
+    def _cancellable_cb():
+        log.debug("sync_call_nm_client[%s]: cancelled", info)
+        return False
+
+    cancellable = Cancellable()
+    cancellable_id = cancellable.connect(_cancellable_cb)
+    callbacks.append(lambda: cancellable.disconnect(cancellable_id))
+
+    def _timeout_cb(*args, **kwargs):
+        log.debug("sync_call_nm_client[%s]: timeout", info)
+        result.timeout = True
+        cancellable.cancel()
+
+    timeout_source = timeout_source_new(int(timeout * 1000))
+    timeout_source.set_callback(_timeout_cb)
+    timeout_source.attach(context)
+
+    def _finish_cb(source_object, async_result):
+        log.debug("sync_call_nm_client[%s]: call %s",
+                  info,
+                  async_call_finish.get_symbol())
+        try:
+            result.received_data = async_call_finish(async_result)
+        except Exception as e:  # pylint: disable=broad-except
+            result.error_message = str(e)
+        finally:
+            _stop_loop()
+
+    context.push_thread_default()
+
+    log.debug("sync_call_nm_client[%s]: call", info)
+    try:
+        async_call(
+            *args,
+            *kwargs,
```

Comment:
It should be `**kwargs`, because it is a dictionary. Maybe it is not necessary to support the keyword arguments as well, so feel free to remove `kwargs` from this function and maybe rename `args` to `call_args`, so it is clear what type of arguments are expected.

Reply:
It is a typo, and yes, I'd prefer to remove the kwargs.

Reply:
Done in https://github.com/rhinstaller/anaconda/pull/4322/commits/1657ee61491554f695d75864c3da51582e762c6c